File: solutions/zig/pythagorean-triplet/1/pythagorean_triplet.zig
const std = @import("std");
const mem = std.mem;

pub const Triplet = struct {
    a: usize,
    b: usize,
    c: usize,

    pub fn init(a: usize, b: usize, c: usize) Triplet {
        return Triplet{ .a = a, .b = b, .c = c };
    }
};

pub fn tripletsWithSum(allocator: mem.Allocator, n: usize) mem.Allocator.Error![]Triplet {
    var triplets = std.ArrayList(Triplet).init(allocator);
    for (3..n / 3) |a| {
        const b_plus_c = n - a;
        const b = (b_plus_c * b_plus_c - a * a) / (2 * b_plus_c);
        const c = n - a - b;
        if (a < b and a * a + b * b == c * c) {
            try triplets.append(Triplet.init(a, b, c));
        }
    }
    return triplets.toOwnedSlice();
}

File: solutions/swift/pythagorean-triplet/1/Sources/PythagoreanTriplet/PythagoreanTriplet.swift
func tripletsWithSum(_ sum: Int) -> [[Int]] {
  var result: [[Int]] = []
  for a in 3..<sum / 3 {
    let b_plus_c = sum - a
    let b = (b_plus_c * b_plus_c - a * a) / (2 * b_plus_c)
    let c = sum - a - b
    if a < b && a * a + b * b == c * c { result.append([a, b, c]) }
  }
  return result
}
